### PR TITLE
fix(shared): avoid indexing unindexable values in vim.tbl_get()

### PR DESCRIPTION
**Pull request by williamboman**
I just saw #18329 was resolved in #18332, but figured I'd open this anyway because it targets `vim.tbl_get()` specifically.

Not sure if throwing a Lua error in these cases is WAI, but I'd argue most people would assume it would not throw but simply return `nil`.

### CODE REVIEW

File: runtime/lua/vim/shared.lua [hunk 1]
```diff
@@ -365,7 +365,10 @@ function vim.tbl_get(o, ...)
   if #keys == 0 then
     return
   end
-  for _, k in ipairs(keys) do
+  for i, k in ipairs(keys) do
+    if type(o[k]) ~= 'table' and next(keys, i) then
```
Comment (williamboman) [inline runtime/lua/vim/shared.lua:369]: The logic here is basically to do an early `return nil` if the current unwrapped value is not a table (i.e. we can't go deeper) _and_ there are more keys to iterate.

**Comment by mfussenegger** [comment]
I think in the linked case it was good that it failed with an error. If tbl_get returned null it would've silently broken the implementation,  making it harder to notice and track down the issue.

I think if there is a non-table in the path it's almost always an error. I wouldn't change it.

**Comment by williamboman** [comment]
> _Edit_: Had another look and I guess it would've worked out fine in that case. But still, if there is a non-table in the path it kinda implies that the programmer made an assumption that was wrong. Not failing could mask bugs.

Yeah I feel like there was an assumption in that implementation that `vim.tbl_get()` would return `nil` in the case it runs into a boolean value. I kinda fail to see its purpose if you have to be conscious about the structure of the table you're accessing? Then why not simply use normal dot notation? I think the primary use case is when dealing with polymorphic table structures, allowing you to access potential values at certain paths. The fact that it craps the bed when it runs into a `number`, `boolean`, `thread`, `function` is a bit unclear imo (why not do the same when it encounters `strings` then?).

I was about to (unknowingly) run into the exact same issue as #18329 because I was intending on using `vim.tbl_get()` for very similar reasons.

**Comment by mfussenegger** [comment]
There are more issues that would be solved by this: https://github.com/neovim/neovim/issues/18344

> Then why not simply use normal dot notation? 

To avoid `(x or {})` nesting to handle the `nil` cases.

I kinda agree that it makes sense for the LSP use cases where there are lots of `bool | somethingElse` structures. I'm still not entirely sure if it's the right thing for the general purpose function, but I'll probably go ahead and merge this. Just giving some time for others to have a look

**Review by gpanders — APPROVED** [review]

**Comment by williamboman** [comment]
Thanks for merging! I just realized this could probably have been cleaner like so 😬 :

```diff
diff --git a/runtime/lua/vim/shared.lua b/runtime/lua/vim/shared.lua
index 172fac3a8..800562377 100644
--- a/runtime/lua/vim/shared.lua
+++ b/runtime/lua/vim/shared.lua
@@ -366,12 +366,11 @@ function vim.tbl_get(o, ...)
     return
   end
   for i, k in ipairs(keys) do
-    if type(o[k]) ~= 'table' and next(keys, i) then
-      return nil
-    end
     o = o[k]
     if o == nil then
       return
+    elseif type(o) ~= 'table' and next(keys, i) then
+      return
     end
   end
   return o
```

**Pull request merged**